File: python/msgpack_blob/_decode.py
"""Internal: decoding & inspection (mirrors msgpack_blob_decode.cpp)."""

from __future__ import annotations

import struct
from typing import Optional, Tuple

from . import _format as F
from .value import Type, Value

__all__ = [
    "RC_OK",
    "RC_ERROR",
    "RC_NOTFOUND",
    "is_valid",
    "error_position",
    "get_type",
    "get_type_str",
    "get_container_count",
    "path_step",
    "lookup",
    "decode_element",
]

RC_OK = 0
RC_ERROR = 1
RC_NOTFOUND = 2

_I64_MAX = (1 << 63) - 1


def is_valid(a, n: int) -> bool:
    if n == 0:
        return False
    return F.skip_one(a, n, 0) == n


def error_position(a, n: int) -> int:
    if n == 0:
        return 0
    if F.skip_one(a, n, 0) == n:
        return 0
    i = 0
    while i < n:
        nxt = F.skip_one(a, n, i)
        if not nxt:
            return i
        i = nxt
    return 0


def _is_timestamp_ext(a, n: int, i: int) -> bool:
    if i >= n:
        return False
    b = a[i]
    if b == F.MP_FIXEXT4 and i + 6 <= n and a[i + 1] == F.MP_TIMESTAMP_TYPE:
        return True
    if b == F.MP_FIXEXT8 and i + 10 <= n and a[i + 1] == F.MP_TIMESTAMP_TYPE:
        return True
    if b == F.MP_EXT8 and i + 3 <= n and a[i + 1] == 12 and a[i + 2] == F.MP_TIMESTAMP_TYPE:
        return True
    return False


def _decode_timestamp(a, n: int, i: int) -> Optional[Tuple[int, int]]:
    if i >= n:
        return None
    b = a[i]
    if b == F.MP_FIXEXT4 and i + 6 <= n and a[i + 1] == F.MP_TIMESTAMP_TYPE:
        return (F.read32(a, i + 2), 0)
    if b == F.MP_FIXEXT8 and i + 10 <= n and a[i + 1] == F.MP_TIMESTAMP_TYPE:
        v = F.read64(a, i + 2)
        return (v & 0x3FFFFFFFF, v >> 34)
    if b == F.MP_EXT8 and i + 15 <= n and a[i + 1] == 12 and a[i + 2] == F.MP_TIMESTAMP_TYPE:
        nsec = F.read32(a, i + 3)
        sec = F.read64(a, i + 7)
        if sec >= (1 << 63):
            sec -= 1 << 64
        return (sec, nsec)
    return None


def get_type(a, n: int, i: int) -> Type:
    if i >= n:
        return Type.NIL
    b = a[i]
    if b == F.MP_NIL:
        return Type.NIL
    if b == F.MP_TRUE:
        return Type.TRUE
    if b == F.MP_FALSE:
        return Type.FALSE
    if b <= 0x7F or b >= 0xE0:
        return Type.INTEGER
    if 0xA0 <= b <= 0xBF:
        return Type.STRING
    if 0x90 <= b <= 0x9F:
        return Type.ARRAY
    if 0x80 <= b <= 0x8F:
        return Type.MAP
    if b in (F.MP_UINT8, F.MP_UINT16, F.MP_UINT32, F.MP_UINT64,
             F.MP_INT8, F.MP_INT16, F.MP_INT32, F.MP_INT64):
        return Type.INTEGER
    if b == F.MP_FLOAT32:
        return Type.FLOAT32
    if b == F.MP_FLOAT64:
        return Type.REAL
    if b in (F.MP_STR8, F.MP_STR16, F.MP_STR32):
        return Type.STRING
    if b in (F.MP_BIN8, F.MP_BIN16, F.MP_BIN32):
        return Type.BINARY
    if b in (F.MP_ARRAY16, F.MP_ARRAY32):
        return Type.ARRAY
    if b in (F.MP_MAP16, F.MP_MAP32):
        return Type.MAP
    if b in (F.MP_EXT8, F.MP_EXT16, F.MP_EXT32, F.MP_FIXEXT1, F.MP_FIXEXT2,
             F.MP_FIXEXT4, F.MP_FIXEXT8, F.MP_FIXEXT16):
        return Type.TIMESTAMP if _is_timestamp_ext(a, n, i) else Type.EXT
    return Type.NIL


def get_type_str(a, n: int, i: int) -> str:
    return get_type(a, n, i).value


def get_container_count(a, n: int, i: int) -> int:
    if i >= n:
        return -1
    b = a[i]
    if 0x90 <= b <= 0x9F:
        return b & 0x0F
    if 0x80 <= b <= 0x8F:
        return b & 0x0F
    if b == F.MP_ARRAY16 and i + 3 <= n:
        return F.read16(a, i + 1)
    if b == F.MP_ARRAY32 and i + 5 <= n:
        return F.read32(a, i + 1)
    if b == F.MP_MAP16 and i + 3 <= n:
        return F.read16(a, i + 1)
    if b == F.MP_MAP32 and i + 5 <= n:
        return F.read32(a, i + 1)
    return -1


def path_step(zpath: str, pi: int):
    """Parse one step of $.key[idx] syntax.

    Returns ``(kind, new_pi, key, idx)`` where *kind* is ``0`` (end),
    ``-1`` (error), ``'k'`` (key) or ``'i'`` (index).
    """
    i = pi
    if i >= len(zpath):
        return (0, i, None, 0)
    c = zpath[i]
    if c == ".":
        i += 1
        start = i
        while i < len(zpath) and zpath[i] not in ".[":
            i += 1
        return ("k", i, zpath[start:i], 0)
    if c == "[":
        idx = 0
        has_digit = False
        i += 1
        while i < len(zpath) and "0" <= zpath[i] <= "9":
            idx = idx * 10 + (ord(zpath[i]) - 48)
            i += 1
            has_digit = True
        if not has_digit or i >= len(zpath) or zpath[i] != "]":
            return (-1, i, None, 0)
        i += 1
        return ("i", i, None, idx)
    return (-1, i, None, 0)


def _key_at(a, n: int, i: int):
    """Return (key_bytes, value_offset) for a map key at *i*, or (None, off)."""
    kb = a[i]
    if 0xA0 <= kb <= 0xBF:
        klen = kb & 0x1F
        koff = i + 1
    elif kb == F.MP_STR8 and i + 2 <= n:
        klen = a[i + 1]
        koff = i + 2
    elif kb == F.MP_STR16 and i + 3 <= n:
        klen = F.read16(a, i + 1)
        koff = i + 3
    elif kb == F.MP_STR32 and i + 5 <= n:
        klen = F.read32(a, i + 1)
        koff = i + 5
    else:
        return None
    if klen > n - koff:
        return None
    return bytes(a[koff:koff + klen])


def lookup(a, n: int, iroot: int, zpath: str):
    """Resolve *zpath* to a byte range. Returns ``(rc, iStart, iEnd)``."""
    if not zpath or zpath[0] != "$":
        return (RC_ERROR, 0, 0)
    icur = iroot
    pi = 1

    while True:
        kind, pi, key, idx = path_step(zpath, pi)

        if kind == 0:
            inext = F.skip_one(a, n, icur)
            istart = icur
            iend = inext if inext else n
            return ((RC_OK if (inext or icur == n) else RC_ERROR), istart, iend)
        if kind == -1:
            return (RC_ERROR, 0, 0)
        if icur >= n:
            return (RC_NOTFOUND, 0, 0)

        if kind == "i":
            b = a[icur]
            if 0x90 <= b <= 0x9F:
                count = b & 0x0F
                elem_off = icur + 1
            elif b == F.MP_ARRAY16:
                if icur + 3 > n:
                    return (RC_ERROR, 0, 0)
                count = F.read16(a, icur + 1)
                elem_off = icur + 3
            elif b == F.MP_ARRAY32:
                if icur + 5 > n:
                    return (RC_ERROR, 0, 0)
                count = F.read32(a, icur + 1)
                elem_off = icur + 5
            else:
                return (RC_NOTFOUND, 0, 0)
            if idx < 0 or idx >= count:
                return (RC_NOTFOUND, 0, 0)
            icur = elem_off
            for _ in range(idx):
                icur = F.skip_one(a, n, icur)
                if not icur:
                    return (RC_ERROR, 0, 0)
        else:
            b = a[icur]
            if 0x80 <= b <= 0x8F:
                count = b & 0x0F
                elem_off = icur + 1
            elif b == F.MP_MAP16:
                if icur + 3 > n:
                    return (RC_ERROR, 0, 0)
                count = F.read16(a, icur + 1)
                elem_off = icur + 3
            elif b == F.MP_MAP32:
                if icur + 5 > n:
                    return (RC_ERROR, 0, 0)
                count = F.read32(a, icur + 1)
                elem_off = icur + 5
            else:
                return (RC_NOTFOUND, 0, 0)
            key_bytes = key.encode("utf-8", "surrogateescape")
            icur = elem_off
            found = False
            j = 0
            while j < count and not found:
                if icur >= n:
                    return (RC_ERROR, 0, 0)
                kstr = _key_at(a, n, icur)
                val_off = F.skip_one(a, n, icur)
                if not val_off:
                    return (RC_ERROR, 0, 0)
                if kstr is not None and kstr == key_bytes:
                    icur = val_off
                    found = True
                else:
                    icur = F.skip_one(a, n, val_off)
                    if not icur:
                        return (RC_ERROR, 0, 0)
                j += 1
            if not found:
                return (RC_NOTFOUND, 0, 0)


def decode_element(a, n: int, istart: int, iend: int) -> Value:
    if istart >= n or istart >= iend:
        return Value.nil()
    b = a[istart]

    if b == F.MP_NIL:
        return Value.nil()
    if b == F.MP_FALSE:
        return Value.boolean(False)
    if b == F.MP_TRUE:
        return Value.boolean(True)
    if b <= 0x7F:
        return Value.integer(b)
    if b >= 0xE0:
        return Value.integer(b - 256)

    if b == F.MP_UINT8:
        if istart + 2 <= n:
            return Value.integer(a[istart + 1])
    elif b == F.MP_UINT16:
        if istart + 3 <= n:
            return Value.integer(F.read16(a, istart + 1))
    elif b == F.MP_UINT32:
        if istart + 5 <= n:
            return Value.integer(F.read32(a, istart + 1))
    elif b == F.MP_UINT64:
        if istart + 9 <= n:
            return Value.unsigned_integer(F.read64(a, istart + 1))
    elif b == F.MP_INT8:
        if istart + 2 <= n:
            v = a[istart + 1]
            return Value.integer(v - 256 if v >= 128 else v)
    elif b == F.MP_INT16:
        if istart + 3 <= n:
            v = F.read16(a, istart + 1)
            return Value.integer(v - (1 << 16) if v >= (1 << 15) else v)
    elif b == F.MP_INT32:
        if istart + 5 <= n:
            v = F.read32(a, istart + 1)
            return Value.integer(v - (1 << 32) if v >= (1 << 31) else v)
    elif b == F.MP_INT64:
        if istart + 9 <= n:
            v = F.read64(a, istart + 1)
            return Value.integer(v - (1 << 64) if v >= (1 << 63) else v)
    elif b == F.MP_FLOAT32:
        if istart + 5 <= n:
            f = struct.unpack(">f", bytes(a[istart + 1:istart + 5]))[0]
            return Value.real32(f)
    elif b == F.MP_FLOAT64:
        if istart + 9 <= n:
            d = struct.unpack(">d", bytes(a[istart + 1:istart + 9]))[0]
            return Value.real(d)

    # str → String
    soff = 0
    slen = 0
    if 0xA0 <= b <= 0xBF:
        slen = b & 0x1F
        soff = istart + 1
    elif b == F.MP_STR8 and istart + 2 <= n:
        slen = a[istart + 1]
        soff = istart + 2
    elif b == F.MP_STR16 and istart + 3 <= n:
        slen = F.read16(a, istart + 1)
        soff = istart + 3
    elif b == F.MP_STR32 and istart + 5 <= n:
        slen = F.read32(a, istart + 1)
        soff = istart + 5
    if soff:
        if slen > n - soff:
            slen = n - soff
        return Value.string(bytes(a[soff:soff + slen]))

    # bin → Binary (payload only)
    boff = 0
    blen = 0
    if b == F.MP_BIN8 and istart + 2 <= n:
        blen = a[istart + 1]
        boff = istart + 2
    elif b == F.MP_BIN16 and istart + 3 <= n:
        blen = F.read16(a, istart + 1)
        boff = istart + 3
    elif b == F.MP_BIN32 and istart + 5 <= n:
        blen = F.read32(a, istart + 1)
        boff = istart + 5
    if boff:
        if blen > n - boff:
            blen = n - boff
        return Value.binary(bytes(a[boff:boff + blen]))

    # timestamp ext
    ts = _decode_timestamp(a, n, istart)
    if ts is not None:
        return Value.timestamp(ts[0], ts[1])

    # ext → Ext (type code + payload)
    tc = 0
    elen = 0
    eoff = 0
    if b == F.MP_FIXEXT1 and istart + 3 <= n:
        tc = a[istart + 1]; elen = 1; eoff = istart + 2
    elif b == F.MP_FIXEXT2 and istart + 4 <= n:
        tc = a[istart + 1]; elen = 2; eoff = istart + 2
    elif b == F.MP_FIXEXT4 and istart + 6 <= n:
        tc = a[istart + 1]; elen = 4; eoff = istart + 2
    elif b == F.MP_FIXEXT8 and istart + 10 <= n:
        tc = a[istart + 1]; elen = 8; eoff = istart + 2
    elif b == F.MP_FIXEXT16 and istart + 18 <= n:
        tc = a[istart + 1]; elen = 16; eoff = istart + 2
    elif b == F.MP_EXT8 and istart + 3 <= n:
        elen = a[istart + 1]; tc = a[istart + 2]; eoff = istart + 3
    elif b == F.MP_EXT16 and istart + 4 <= n:
        elen = F.read16(a, istart + 1); tc = a[istart + 3]; eoff = istart + 4
    elif b == F.MP_EXT32 and istart + 6 <= n:
        elen = F.read32(a, istart + 1); tc = a[istart + 5]; eoff = istart + 6
    if eoff:
        if elen > n - eoff:
            elen = n - eoff
        tc_signed = tc - 256 if tc >= 128 else tc
        return Value.ext(tc_signed, bytes(a[eoff:eoff + elen]))

    # containers → raw binary blob (includes header)
    return Value.binary(bytes(a[istart:iend]))

File: rust/src/decode.rs
//! Internal: decoding & inspection (mirrors `msgpack_blob_decode.cpp`).

use crate::format as f;
use crate::value::{Type, Value};

pub const RC_OK: i32 = 0;
pub const RC_ERROR: i32 = 1;
pub const RC_NOTFOUND: i32 = 2;

pub fn is_valid(a: &[u8], n: usize) -> bool {
    if n == 0 {
        return false;
    }
    f::skip_one(a, n, 0) == n
}

pub fn error_position(a: &[u8], n: usize) -> usize {
    if n == 0 {
        return 0;
    }
    if f::skip_one(a, n, 0) == n {
        return 0;
    }
    let mut i = 0;
    while i < n {
        let nxt = f::skip_one(a, n, i);
        if nxt == 0 {
            return i;
        }
        i = nxt;
    }
    0
}

fn is_timestamp_ext(a: &[u8], n: usize, i: usize) -> bool {
    if i >= n {
        return false;
    }
    let b = a[i];
    if b == f::MP_FIXEXT4 && i + 6 <= n && a[i + 1] == f::MP_TIMESTAMP_TYPE {
        return true;
    }
    if b == f::MP_FIXEXT8 && i + 10 <= n && a[i + 1] == f::MP_TIMESTAMP_TYPE {
        return true;
    }
    if b == f::MP_EXT8 && i + 3 <= n && a[i + 1] == 12 && a[i + 2] == f::MP_TIMESTAMP_TYPE {
        return true;
    }
    false
}

fn decode_timestamp(a: &[u8], n: usize, i: usize) -> Option<(i64, u32)> {
    if i >= n {
        return None;
    }
    let b = a[i];
    if b == f::MP_FIXEXT4 && i + 6 <= n && a[i + 1] == f::MP_TIMESTAMP_TYPE {
        return Some((f::read32(a, i + 2) as i64, 0));
    }
    if b == f::MP_FIXEXT8 && i + 10 <= n && a[i + 1] == f::MP_TIMESTAMP_TYPE {
        let v = f::read64(a, i + 2);
        return Some(((v & 0x3_FFFF_FFFF) as i64, (v >> 34) as u32));
    }
    if b == f::MP_EXT8 && i + 15 <= n && a[i + 1] == 12 && a[i + 2] == f::MP_TIMESTAMP_TYPE {
        let nsec = f::read32(a, i + 3);
        let sec = f::read64(a, i + 7) as i64;
        return Some((sec, nsec));
    }
    None
}

pub fn get_type(a: &[u8], n: usize, i: usize) -> Type {
    if i >= n {
        return Type::Nil;
    }
    let b = a[i];
    if b == f::MP_NIL {
        return Type::Nil;
    }
    if b == f::MP_TRUE {
        return Type::True;
    }
    if b == f::MP_FALSE {
        return Type::False;
    }
    if b <= 0x7f || b >= 0xe0 {
        return Type::Integer;
    }
    if (0xa0..=0xbf).contains(&b) {
        return Type::String;
    }
    if (0x90..=0x9f).contains(&b) {
        return Type::Array;
    }
    if (0x80..=0x8f).contains(&b) {
        return Type::Map;
    }
    match b {
        f::MP_UINT8
        | f::MP_UINT16
        | f::MP_UINT32
        | f::MP_UINT64
        | f::MP_INT8
        | f::MP_INT16
        | f::MP_INT32
        | f::MP_INT64 => Type::Integer,
        f::MP_FLOAT32 => Type::Float32,
        f::MP_FLOAT64 => Type::Real,
        f::MP_STR8 | f::MP_STR16 | f::MP_STR32 => Type::String,
        f::MP_BIN8 | f::MP_BIN16 | f::MP_BIN32 => Type::Binary,
        f::MP_ARRAY16 | f::MP_ARRAY32 => Type::Array,
        f::MP_MAP16 | f::MP_MAP32 => Type::Map,
        f::MP_EXT8
        | f::MP_EXT16
        | f::MP_EXT32
        | f::MP_FIXEXT1
        | f::MP_FIXEXT2
        | f::MP_FIXEXT4
        | f::MP_FIXEXT8
        | f::MP_FIXEXT16 => {
            if is_timestamp_ext(a, n, i) {
                Type::Timestamp
            } else {
                Type::Ext
            }
        }
        _ => Type::Nil,
    }
}

pub fn get_container_count(a: &[u8], n: usize, i: usize) -> i64 {
    if i >= n {
        return -1;
    }
    let b = a[i];
    if (0x90..=0x9f).contains(&b) {
        return (b & 0x0f) as i64;
    }
    if (0x80..=0x8f).contains(&b) {
        return (b & 0x0f) as i64;
    }
    if b == f::MP_ARRAY16 && i + 3 <= n {
        return f::read16(a, i + 1) as i64;
    }
    if b == f::MP_ARRAY32 && i + 5 <= n {
        return f::read32(a, i + 1) as i64;
    }
    if b == f::MP_MAP16 && i + 3 <= n {
        return f::read16(a, i + 1) as i64;
    }
    if b == f::MP_MAP32 && i + 5 <= n {
        return f::read32(a, i + 1) as i64;
    }
    -1
}

/// One parsed step of `$.key[idx]` syntax.
pub enum Step<'a> {
    End,
    Error,
    Key(&'a str),
    Index(i64),
}

/// Parse one step starting at byte index `pi` in `zpath`; returns the step and
/// the next index.
pub fn path_step(zpath: &[u8], pi: usize) -> (Step<'_>, usize) {
    let mut i = pi;
    if i >= zpath.len() {
        return (Step::End, i);
    }
    let c = zpath[i];
    if c == b'.' {
        i += 1;
        let start = i;
        while i < zpath.len() && zpath[i] != b'.' && zpath[i] != b'[' {
            i += 1;
        }
        // Path key bytes are ASCII/UTF-8 from the caller's &str.
        let key = std::str::from_utf8(&zpath[start..i]).unwrap_or("");
        return (Step::Key(key), i);
    }
    if c == b'[' {
        let mut idx: i64 = 0;
        let mut has_digit = false;
        i += 1;
        while i < zpath.len() && zpath[i].is_ascii_digit() {
            idx = idx * 10 + (zpath[i] - b'0') as i64;
            i += 1;
            has_digit = true;
        }
        if !has_digit || i >= zpath.len() || zpath[i] != b']' {
            return (Step::Error, i);
        }
        i += 1;
        return (Step::Index(idx), i);
    }
    (Step::Error, i)
}

/// Key bytes of a map key at `i`, or `None` if not a string key.
fn key_at(a: &[u8], n: usize, i: usize) -> Option<&[u8]> {
    let kb = a[i];
    let (klen, koff) = if (0xa0..=0xbf).contains(&kb) {
        ((kb & 0x1f) as usize, i + 1)
    } else if kb == f::MP_STR8 && i + 2 <= n {
        (a[i + 1] as usize, i + 2)
    } else if kb == f::MP_STR16 && i + 3 <= n {
        (f::read16(a, i + 1) as usize, i + 3)
    } else if kb == f::MP_STR32 && i + 5 <= n {
        (f::read32(a, i + 1) as usize, i + 5)
    } else {
        return None;
    };
    if klen > n - koff {
        return None;
    }
    Some(&a[koff..koff + klen])
}

/// Resolve `zpath` to a byte range. Returns `(rc, i_start, i_end)`.
pub fn lookup(a: &[u8], n: usize, iroot: usize, zpath: &str) -> (i32, usize, usize) {
    let zb = zpath.as_bytes();
    if zb.is_empty() || zb[0] != b'$' {
        return (RC_ERROR, 0, 0);
    }
    let mut icur = iroot;
    let mut pi = 1;

    loop {
        let (step, npi) = path_step(zb, pi);
        pi = npi;
        match step {
            Step::End => {
                let inext = f::skip_one(a, n, icur);
                let iend = if inext != 0 { inext } else { n };
                let rc = if inext != 0 || icur == n {
                    RC_OK
                } else {
                    RC_ERROR
                };
                return (rc, icur, iend);
            }
            Step::Error => return (RC_ERROR, 0, 0),
            Step::Index(idx) => {
                if icur >= n {
                    return (RC_NOTFOUND, 0, 0);
                }
                let b = a[icur];
                let (count, elem_off) = if (0x90..=0x9f).contains(&b) {
                    ((b & 0x0f) as i64, icur + 1)
                } else if b == f::MP_ARRAY16 {
                    if icur + 3 > n {
                        return (RC_ERROR, 0, 0);
                    }
                    (f::read16(a, icur + 1) as i64, icur + 3)
                } else if b == f::MP_ARRAY32 {
                    if icur + 5 > n {
                        return (RC_ERROR, 0, 0);
                    }
                    (f::read32(a, icur + 1) as i64, icur + 5)
                } else {
                    return (RC_NOTFOUND, 0, 0);
                };
                if idx < 0 || idx >= count {
                    return (RC_NOTFOUND, 0, 0);
                }
                icur = elem_off;
                for _ in 0..idx {
                    icur = f::skip_one(a, n, icur);
                    if icur == 0 {
                        return (RC_ERROR, 0, 0);
                    }
                }
            }
            Step::Key(key) => {
                if icur >= n {
                    return (RC_NOTFOUND, 0, 0);
                }
                let b = a[icur];
                let (count, elem_off) = if (0x80..=0x8f).contains(&b) {
                    ((b & 0x0f) as usize, icur + 1)
                } else if b == f::MP_MAP16 {
                    if icur + 3 > n {
                        return (RC_ERROR, 0, 0);
                    }
                    (f::read16(a, icur + 1) as usize, icur + 3)
                } else if b == f::MP_MAP32 {
                    if icur + 5 > n {
                        return (RC_ERROR, 0, 0);
                    }
                    (f::read32(a, icur + 1) as usize, icur + 5)
                } else {
                    return (RC_NOTFOUND, 0, 0);
                };
                let key_bytes = key.as_bytes();
                icur = elem_off;
                let mut found = false;
                let mut j = 0;
                while j < count && !found {
                    if icur >= n {
                        return (RC_ERROR, 0, 0);
                    }
                    let kstr = key_at(a, n, icur);
                    let val_off = f::skip_one(a, n, icur);
                    if val_off == 0 {
                        return (RC_ERROR, 0, 0);
                    }
                    if kstr == Some(key_bytes) {
                        icur = val_off;
                        found = true;
                    } else {
                        icur = f::skip_one(a, n, val_off);
                        if icur == 0 {
                            return (RC_ERROR, 0, 0);
                        }
                    }
                    j += 1;
                }
                if !found {
                    return (RC_NOTFOUND, 0, 0);
                }
            }
        }
    }
}

pub fn decode_element(a: &[u8], n: usize, istart: usize, iend: usize) -> Value {
    if istart >= n || istart >= iend {
        return Value::nil();
    }
    let b = a[istart];

    if b == f::MP_NIL {
        return Value::nil();
    }
    if b == f::MP_FALSE {
        return Value::boolean(false);
    }
    if b == f::MP_TRUE {
        return Value::boolean(true);
    }
    if b <= 0x7f {
        return Value::integer(b as i64);
    }
    if b >= 0xe0 {
        return Value::integer(b as i8 as i64);
    }

    match b {
        f::MP_UINT8 => {
            if istart + 2 <= n {
                return Value::integer(a[istart + 1] as i64);
            }
        }
        f::MP_UINT16 => {
            if istart + 3 <= n {
                return Value::integer(f::read16(a, istart + 1) as i64);
            }
        }
        f::MP_UINT32 => {
            if istart + 5 <= n {
                return Value::integer(f::read32(a, istart + 1) as i64);
            }
        }
        f::MP_UINT64 => {
            if istart + 9 <= n {
                return Value::unsigned_integer(f::read64(a, istart + 1));
            }
        }
        f::MP_INT8 => {
            if istart + 2 <= n {
                return Value::integer(a[istart + 1] as i8 as i64);
            }
        }
        f::MP_INT16 => {
            if istart + 3 <= n {
                return Value::integer(f::read16(a, istart + 1) as u16 as i16 as i64);
            }
        }
        f::MP_INT32 => {
            if istart + 5 <= n {
                return Value::integer(f::read32(a, istart + 1) as i32 as i64);
            }
        }
        f::MP_INT64 => {
            if istart + 9 <= n {
                return Value::integer(f::read64(a, istart + 1) as i64);
            }
        }
        f::MP_FLOAT32 => {
            if istart + 5 <= n {
                let bits = f::read32(a, istart + 1);
                return Value::real32(f32::from_bits(bits));
            }
        }
        f::MP_FLOAT64 => {
            if istart + 9 <= n {
                let bits = f::read64(a, istart + 1);
                return Value::real(f64::from_bits(bits));
            }
        }
        _ => {}
    }

    // str
    let (mut slen, soff) = if (0xa0..=0xbf).contains(&b) {
        ((b & 0x1f) as usize, istart + 1)
    } else if b == f::MP_STR8 && istart + 2 <= n {
        (a[istart + 1] as usize, istart + 2)
    } else if b == f::MP_STR16 && istart + 3 <= n {
        (f::read16(a, istart + 1) as usize, istart + 3)
    } else if b == f::MP_STR32 && istart + 5 <= n {
        (f::read32(a, istart + 1) as usize, istart + 5)
    } else {
        (0, 0)
    };
    if soff != 0 {
        if slen > n - soff {
            slen = n - soff;
        }
        return Value::string_bytes(&a[soff..soff + slen]);
    }

    // bin
    let (mut blen, boff) = if b == f::MP_BIN8 && istart + 2 <= n {
        (a[istart + 1] as usize, istart + 2)
    } else if b == f::MP_BIN16 && istart + 3 <= n {
        (f::read16(a, istart + 1) as usize, istart + 3)
    } else if b == f::MP_BIN32 && istart + 5 <= n {
        (f::read32(a, istart + 1) as usize, istart + 5)
    } else {
        (0, 0)
    };
    if boff != 0 {
        if blen > n - boff {
            blen = n - boff;
        }
        return Value::binary(&a[boff..boff + blen]);
    }

    // timestamp
    if let Some((sec, nsec)) = decode_timestamp(a, n, istart) {
        return Value::timestamp_ns(sec, nsec);
    }

    // ext
    let (tc, mut elen, eoff): (i8, usize, usize) = match b {
        f::MP_FIXEXT1 if istart + 3 <= n => (a[istart + 1] as i8, 1, istart + 2),
        f::MP_FIXEXT2 if istart + 4 <= n => (a[istart + 1] as i8, 2, istart + 2),
        f::MP_FIXEXT4 if istart + 6 <= n => (a[istart + 1] as i8, 4, istart + 2),
        f::MP_FIXEXT8 if istart + 10 <= n => (a[istart + 1] as i8, 8, istart + 2),
        f::MP_FIXEXT16 if istart + 18 <= n => (a[istart + 1] as i8, 16, istart + 2),
        f::MP_EXT8 if istart + 3 <= n => (a[istart + 2] as i8, a[istart + 1] as usize, istart + 3),
        f::MP_EXT16 if istart + 4 <= n => (
            a[istart + 3] as i8,
            f::read16(a, istart + 1) as usize,
            istart + 4,
        ),
        f::MP_EXT32 if istart + 6 <= n => (
            a[istart + 5] as i8,
            f::read32(a, istart + 1) as usize,
            istart + 6,
        ),
        _ => (0, 0, 0),
    };
    if eoff != 0 {
        if elen > n - eoff {
            elen = n - eoff;
        }
        return Value::ext(tc, &a[eoff..eoff + elen]);
    }

    // containers → raw binary blob (includes header)
    Value::binary(&a[istart..iend])
}
